### PR TITLE
Add `('a, 'b)  Either.t = Left of 'a | Right of 'b` and `List.partition_map`

### PR DESCRIPTION
**Pull request by gasche**
This is a new attempt at #9047 (introducing `List.partition_map : ('a -> ('b, 'c) result) -> 'a list -> 'b list * 'c list`) with a brand new `Either.t` type.

(cc @yallop, @nojb)

```ocaml
val left : 'a -> ('a, 'b) t
val right : 'b -> ('a, 'b) t
val find_left : ('a, 'b) t -> 'a option
val find_right : ('a, 'b) t -> 'b option
val swap : ('a, 'b) t -> ('b, 'a) t
val map_left : ('a1 -> 'a2) -> ('a1, 'b) t -> ('a2, 'b) t
val map_right : ('b1 -> 'b2) -> ('a, 'b1) t -> ('a, 'b2) t
val map_both :
  left:('a1 -> 'a2) -> right:('b1 -> 'b2) -> ('a1, 'b1) t -> ('a2, 'b2) t
val fold : left:('a -> 'c) -> right:('b -> 'c) -> ('a, 'b) t -> 'c
val equal :
  left:('a -> 'a -> bool) -> right:('b -> 'b -> bool) ->
  ('a, 'b) t -> ('a, 'b) t -> bool
val compare :
  left:('a -> 'a -> int) -> right:('b -> 'b -> int) ->
  ('a, 'b) t -> ('a, 'b) t -> int
```

Unlike [result], no [either] type is made available in Stdlib,
one needs to access [Either.t] explicitly:

- This type is less common in typical OCaml codebases,
  which prefer domain-specific variant types whose constructors
  carry more meaning.
- Adding this to Stdlib would raise warnings in existing codebases
  that already use a constructor named Left or Right:
  + when opening a module that exports such a name,
    warning 45 is raised
  + adding a second constructor of the same name in scope kicks
    in the disambiguation mechanisms, and warning 41 may now
    be raised by existing code.

If the use becomes more common in the future we can always
revisit this choice.

### CODE REVIEW

File: stdlib/either.mli
Comment (trefis) [inline stdlib/either.mli]: :roll_eyes:

File: stdlib/either.mli
Comment (nojb) [inline stdlib/either.mli]: s/Parsial/Parsifal ?

File: stdlib/either.mli
Comment (trefis) [inline stdlib/either.mli]: Why `4.11` here and `NEXT_RELEASE` in `list.mli`?

File: Changes
Comment (trefis) [inline Changes]: This entry is present twice.

**Review by trefis — COMMENTED** [review]
I believe you forgot to commit `either.ml`.

**Review by nojb — COMMENTED** [review]
We could also add `partition_map` to the `Map` module.

File: stdlib/either.mli
Comment (dbuenzli) [inline stdlib/either.mli]: The behaviours for these names are inconsistent with `Result.get_{ok,error}` and `Option.get`. `get` is for when you know the thing is here (and raises `Invalid_argument` if it's not). I would like these names with that semantics. Names and signature for what you suggest consistent with `Result` and `Option` could be:
```ocaml
val left_value : ('a, 'b) t -> default:'a -> 'a 
val right_value : ('a, 'b) t -> default:'b -> 'b
```
or if you want to keep the current signatures:
```ocaml
val find_left : ('a, 'b) t -> 'a option 
val find_right : ('a, 'b) t -> 'b option
```

File: stdlib/either.mli
Comment (dbuenzli) [inline stdlib/either.mli]: I would call this `map` alone I think.

File: stdlib/either.mli
Comment (dbuenzli) [inline stdlib/either.mli:93]: It would be nice to add predicates `is_left` and `is_right` to mirror `is_ok` and `is_error`.

File: stdlib/either.mli
Comment (dbuenzli) [inline stdlib/either.mli:92]: Other modules usually have `iter` aswell (though personally I see it as a special case of `fold`).

File: stdlib/either.mli
Comment (gasche) [reply stdlib/either.mli]: Sure, I'll use `find_`. I'm not so interested in the more complex accessors, I think the `option` result is useful to be combined with option-to-{list,seq} transformations.

Comment (lpw25) [reply stdlib/either.mli]: I vote for the `find_left` and `find_right` version.

File: stdlib/either.mli
Comment (gasche) [reply stdlib/either.mli:93]: I started with them (I just copied your API from result), but then I got rid of them. I'm not fully convinced they are better than `(function Left _ -> true | Right _ -> false)`, and I don't want to encourage the style of "if Either.is_left ... then ... else ...", which I find inelegant.

File: stdlib/either.mli
Comment (gasche) [reply stdlib/either.mli:92]: I started with it, but it was just an alias for `fold` so I wasn't sure what to do. Maybe I'll add it again. Could we have some extra opinions on this?

Comment (lpw25) [reply stdlib/either.mli:92]: I would like to have an `iter`. I think it conveys the intention better than a `fold` returning `unit`.

File: stdlib/either.mli
Comment (gasche) [reply stdlib/either.mli]: They were written at different times. I think it's fine to ignore this for now.

File: stdlib/either.mli
Comment (lpw25) [reply stdlib/either.mli:93]: As a point of reference: they appear in the corresponding Haskell module.

**Comment by gasche** [comment]
I addressed the first round of review comments (thanks!); I'm going to be away-from-network for most of the afternoon.

File: stdlib/either.mli
Comment (dbuenzli) [reply stdlib/either.mli:93]: @gasche the point is not to write `if Either.is_left ... then ... else ...`. The point is to be able to write `List.filter Either.is_left` instead of `List.filter (function Left _ -> true | Right _ -> false)`.

File: stdlib/either.mli
Comment (dbuenzli) [reply stdlib/either.mli]: I would still like to have the `get_{left,right}` accessors for 1) consistency with `Result` and `Option` 2) the error is slightly more informative than `function Left v -> v | _ -> assert false` when it fails.

File: stdlib/either.mli
Comment (dbuenzli) [reply stdlib/either.mli]: No strong opinion, but we have `Fun.flip` with similar structure, maybe this could be `flip`.

File: stdlib/either.mli
Comment (yallop) [reply stdlib/either.mli:93]: I'm not sure about this: is it useful to build a list of `Left` values (`[Left x; Left y; Left z; ...]`)?  I'd expect `List.filter_map Either.get_left` (which will build `[x; y; z; ...]` instead) to be used more often.

File: stdlib/either.mli
Comment (yallop) [inline stdlib/either.mli:36]: The introduction gives the impression that `result` should be the default, and `Either.t` a kind of fall-back when `result` won't do, which sounds a bit apologetic.  (Similarly below, in the comment starting "Unlike [result]")

File: stdlib/either.mli
Comment (yallop) [inline stdlib/either.mli:72]: I may be alone in this, but I would prefer a function that provides a default for the case where the argument is `Right` rather than returning an option:
```
val from_left : ('a, 'b) t -> 'a -> 'a
```

File: stdlib/either.mli
Comment (dbuenzli) [reply stdlib/either.mli:93]: @yallop yes it is ! Here are few arguments.

1. You may be interested in filtering things before adding more eithers to a list or in general to a filterable data structure.
2. You may want to check everything is left or right but still keep it in either, so you want to write `List.for_all Either.is_left` rather than `List.forall (function Left _ -> true | Right _ -> false)`.
3. We recently introduced `Option`, `Result` whose interfaces were reviewed to death. I think now `Either` can afford consistency with those without us spending days in discussing it (both `Result.is_{ok,error}`, `Option.is_{none,some}` do exist).

File: stdlib/either.mli
Comment (dbuenzli) [reply stdlib/either.mli:72]: I think `Either.from_left` doesn't read right :-). Here again `Result` and `Option` should guide us (see my now outdated comment for a suggestion).

Comment (yallop) [reply stdlib/either.mli:72]: If we let `Result` and `Option` guide us then we should add something like the `value` function in those modules, which is precisely what `from_left` is.  (Furthermore, Haskell has `fromLeft`, too.)

Comment (dbuenzli) [reply stdlib/either.mli:72]: Yes I suggested in the outdated comment (as per `{Result,Option}.value`):

```ocaml
val left_value : ('a, 'b) t -> default:'a -> 'a 
val right_value : ('a, 'b) t -> default:'b -> 'b
```
Note that Haskell doesn't qualify `fromLeft` by `Either` so I don't think we should follow their names on that one, it is confusing.

Comment (yallop) [reply stdlib/either.mli:72]: Ah, it's just the name you're unhappy with.  I don't mind using a different name.  (I had difficulty finding the outdated comment; the only way I can find it was via an email notification.)

File: stdlib/either.mli
Comment (gasche) [reply stdlib/either.mli:92]: Let's add `iter` back.

File: stdlib/either.mli
Comment (gasche) [reply stdlib/either.mli:36]: I think that it makes sense to compare to the type that was introduced earlier.

I tried to give a more positive description of when either should be used, and I had the idea the idea of saying that it makes sense for "symmetric" sums, but actually this is not correct: most use-cases of either will give very asymmetric meanings to the two cases, it's just that the mapping is not a-priori directed by the idea that there is a "good" and "bad" case.

One thing I could say is that either is a primitive sum type whose constructor names are generic/interchangeable, and *then* point out that for values that encode a form of  failure `result` is more precise?

File: stdlib/either.mli
Comment (gasche) [reply stdlib/either.mli:72]: I added `get_left` (now `find_left`) late in my design process. It's the only function in this API that cannot be typed with a linear arrow on the `either` input -- it discards information in some cases -- and I'm not fond of it. The reason to have it is to let people easily get a sequence-like datastructure out of their values.

I personally don't like these proposals that contain a negative encoding of the option type, they have complex type. One should be able to do the exact same thing by operating on the resulting `option` value, with a fairly simple chaining (and more! `to_seq` etc.).

If a consensus emerges that my option-returning selectors are not liked, I will remove them.

File: stdlib/either.mli
Comment (gasche) [reply stdlib/either.mli:93]: I think that `List.filter Either.is_left` is bad design, because it returns a list that you know contains only `'a` (starting from `('a, 'b) Either.t`), but you cannot actually get those `'a` without using an unsafe/partial function. You should use `List.filter_map Either.find_left` instead.

I'm not convinced by `List.for_all Either.is_left` either. The point of having a list of either values is that both cases can occur in it. If you're creating a value that at some point is known by construction to have only `Left (v : 'a)` elements, why not create a list of `'a` directly at this point? If really the way the postcondition holds is too complex to be built into the types in this way, well this sounds infrequent enough that just using `match ... ` or `function ...` is fine.

I'm still looking for a use-case of `is_left` that I personally find convincing. I'm not in favor of adding the function unless one is provided.

(I did not participate much to the review of the `Option` and `Result` type -- and in any case I wanted the types in more than I cared to debate detail points of the API -- but also there I don't think that those are good functions to have. Not my problem, and not my PR.)

Comment (dbuenzli) [reply stdlib/either.mli:93]: Who are you to judge what I want to do with these lists ? Again I may want to keep lefts or rights from from a particular list to compose it with another list of eithers. 

Besides regardless of whether you did participate or not to the `Option` and `Result` PRs, stdlib PR are also judged on consistency of the stdlib. By leaving these functions out you are making it less consistent. 

Frankly I think if you simply had taken these two recent modules as a complete design blueprint (and maybe added `find_{left,right}`) we would already be done at that point.

File: stdlib/either.mli
Comment (yallop) [reply stdlib/either.mli:36]: > I think that it makes sense to compare to the type that was introduced earlier.

I actually disagree with this.  Most OCaml users are in the future, and (if this is merged) will encounter a standard library that already contains both `either` and `result`; there's no reason to present them with documentation that depends on the order in which the types were introduced.

On the other hand, I think it's a good idea to have your justification for `either` in the pull request and the commit message.

Comment (gasche) [reply stdlib/either.mli:36]: I'll try to rewrite this without a reference to `result`.

File: stdlib/either.mli
Comment (gasche) [reply stdlib/either.mli:93]: If you use `List.filter_map Either.find_left`, you can still merge the list with other either values if you wish by making a `List.map Either.left` first. I suspect that in most cases the code will be clearer that way. Of course there are always some odd situations where unnatural code patterns emerge, but I think in this case it's fine to have less stdlib support -- especially if making those unnatural code patterns easier also risks nudging users towards worse decisions in more common cases.

Can you or someone else suggest a *compelling* use-case of these `is_*` functions?

I tried to use github search a bit, and most uses of `Option.is_some` are in code where pattern-matching would be better (see for example [here](https://github.com/ngorogiannis/cyclist/blob/6d4ad7ac5baa7286f376a4726baa8495e393cadd/src/util/tags.ml)). `is_{ok,error}` seems to have some slightly more believable use-cases (here are the [search results for Result.is_error](https://github.com/search?p=2&q=Result.is_error++extension%3Aml&type=Code)), mostly in testing code. But I don't think that these uses (related to the fact that it's typically enough to check if the result is Error to know that a process failed) would port well to Either use-cases.

File: stdlib/either.mli
Comment (gasche) [reply stdlib/either.mli:92]: I also added `for_all` for good measure.

File: stdlib/either.mli
Comment (gasche) [reply stdlib/either.mli]: I changed to use `flip`.

The use-case for this function is when someone choose a particular ordering between Left and Right for an API (for example they provide you with a function from `('a, unit) Either.t` to `'a option`, but not one for `(unit, 'a) Either.t`), and your data naturally comes in the other order.

File: stdlib/either.mli
Comment (dbuenzli) [reply stdlib/either.mli:93]: > Can you or someone else suggest a _compelling_ use-case of these `is_*` functions?

Stream source merges/split/re-source. Test suites. Whatever. Anything that *your* brain may not have imagined of the form: I want to take the lefts of that data structure and put them in the eithers of that other data structure. For that I'm sorry but, `map Either.left (filter_map Either.find_left v)` is *not* clearer than `filter Either.is_left v` (and I won't even talk about allocation this discussion is getting absurd).

I can't believe you are proposing an inconsistent design (including the way you initially called the functions you now renamed to `find_`) just one year after we collectively agreed on one (FWIW I was not ok with all of what when into `Option` and `Result` modules). I think this is just disrespectful for the work that has been done and a waste of (at least my) time.

File: stdlib/either.mli
Comment (gasche) [reply stdlib/either.mli]: done, thanks!

**Comment by gasche** [comment]
I addressed the second round of reviews. Thanks!

**Comment by trefis** [comment]
I find the contrast between:

> Can you or someone else suggest a compelling use-case of these `is_*` functions?

and:

> The use-case for this function is when someone choose a particular ordering between `Left` and `Right` for an API (for example they provide you with a function from `('a, unit) Either.t` to `'a option`, but not one for `(unit, 'a) Either.t`), and your data naturally comes in the other order.

amusing.

I honestly don't really care about the existence (or absence) of `flip`, but on the topic of the `is_*` function I believe @dbuenzli (despite the tone of his reply) makes a good point when talking about consistency.
In particular:
> I can't believe you are proposing an inconsistent design (including the way you initially called the functions you now renamed to find_) just one year after we collectively agreed on one

seems like a fair comment to me.

Looking at these old conversations, I stumbled on https://github.com/ocaml/ocaml/pull/2111#issuecomment-431422570, and I believe the "Design goal: coherence of interfaces" point is relevant here.

**Comment by gasche** [comment]
Different types have different needs, we don't have a `List.is_cons` function either. For `result` it seems like a somewhat idiomatic thing to test whether a computation result is a success or a failure, but on the other hand all use-cases of `is_some` I've seen in the wild are either meh or crappy code. We should remove `is_some`, rather than reproducing this mistake in all future modules.

I personally think that `flip` makes sense (it's the only parametric function that takes and return an either and is not the identity, so it is *bound* to be useful), and I can't see any code smell that is facilitated by having it around. But if you keep making derogatory comments about it, I'll get rid of it :-)

**Comment by dbuenzli** [comment]
> Different types have different needs, we don't have a `List.is_cons` function either.

But `List.is_empty` is a function I often miss...

> I've seen in the wild are either meh or crappy code. We should remove `is_some`, rather than reproducing this mistake in all future modules.

I think you are bit quick to call that a mistake however I don't want to enter into that discussion. But it should be mentioned that making points by searching github is not without bias  1) most apps are not published on github, you mainly get library code 2) The functions were introduced in OCaml 4.08, most libraries keep compatibility with much older ocaml versions which don't have the modules so you likely can't see usage of these for now.

Now speaking of reproducing actual *real*, *known*,  *everyone-agrees* on  mistakes. I cannot not mention that when I suggested we should introduce the *new* entry point `Int.of_string : string -> t option` rather than have it raise an exception, everyone screamed at me about consistency and I had to remove it from my proposal in order to be merged.

So if we want to remain consistent with the recent `Option` and `Result` additions this proposal is still missing:

```ocaml
val get_left : ('a, 'b) t -> 'a (* Raises invalid_argument if it's not a left *)
val get_right : ('a, 'b) t -> 'b (* Raises invalid_argument if it's not a right *)
val left_value : ('a, 'b) t -> default:'a -> 'a 
val right_value : ('a, 'b) t -> default:'b -> 'b
val is_left : ('a, b') -> bool 
val is_right : ('a, 'b) -> bool
```

Note that the last four functions are both in [`base`](https://ocaml.janestreet.com/ocaml-core/latest/doc/base/Base/Either/index.html) and the [haskell stdlib](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Either.html) have them. (It seems both batteries and containers lack an either type). 

The first two are useful when you operate on `either` and end up with an invariant that you cannot or don't want to express via the type system but makes sure your data is in one of the other forms. They are better than simple `Option.get` chaining because breaking the invariant leaks a more informative error message and using them avoids the entirely needless allocation you need by going through `option` if the invariant is not broken. The name and behaviour follows the convention for indexing types in the stdlib e.g. `{Array,String,Bytes}.get`, `Result.get_{ok,error}`, `Option.get`.

**Comment by dbuenzli** [comment]
> We should remove `is_some`, rather than reproducing this mistake in all future modules.

And just for the sake of it here are a few uses of `Option.is_some` from a private application I have (lifted in react signals) that I personally don't find crappy: 

```
> git grep Option.is_some 
dui.ml:    let enabled = S.Bool.(enabled && S.Option.is_some angle) in
dui.ml:    let enabled = S.Option.is_some scale in
dui.ml:    let enabled = S.Option.is_some plan in
dui.ml:    let enabled = S.Option.is_some surf in
dui_editor.ml:  let has_plan = S.Option.is_some plan in
```
Basically these `enabled` boolean signals allow to enable or disable part of the user interface according to whether the data model has some of the information or not.

**Comment by mshinwell** [comment]
I agree with @dbuenzli that we should have the "`is`" functions.  I also often miss `List.is_empty`.

**Comment by gasche** [comment]
Besides `List.partition`, here is another concrete use-case where `('a, 'b) Either.t` is useful. I have two functions that parse and interpret a certain data structure, and both stages can emit errors:

```
type t = ...
type value = ...

type parse_error = ...
val parse : in_channel -> (t, parse_error) result

type eval_error = ...
val eval : t -> (value, eval_error) result
```

Now I may want to provide an API function that composes these two steps, because they are often chained together by my consumers:

```
val parse_and_eval : in_channel -> (value, (parse_error, eval_error) Either.t) result
```

We use `Either.t` to indicate that doing the two operations in sequence may return either a parsing error or an evaluation error. Note that using `result` there would be a bad choice, because `| Ok parse_error -> ...` is very confusing.

**Comment by gasche** [comment]
I have no plan to add `is_{left,right}` in my PR. I agree that consistency is a good thing, but I'm convinced that these functions encourage bad style, and I would rather nudge people into writing good code (or have them define their own helper functions) than automatically transpose helper functions across modules that have fairly different use-cases.

I'm all for compromises and I made various changes in response to the feedback in this PR. (This is a slightly hypocritical comment as I believe that all change suggestions I followed make the API better. If something has to be sacrificed on the altar of compromise, I'm willing to remove the `flip` I am attached to.) But I'm not interested in proposing code that I actively dislike as my own.

Personally I believe that it should be consensual that this module as-is, even *without* the `is_{left,right}` functions, would be a (small) improvement to the standard library. If people start using it and they have further comments to make about the desirability of `is_{left,right}` from experience and real-world use-cases, we can have a discussion about them again.

**Comment by gadmm** [comment]
In this discussion, I think it would be good to remember the advice that one designs for the working programmer.

**Review by dbuenzli — CHANGES_REQUESTED** [review]
Interestingly I have been falling recently on quite a few more use cases of `Option.is_some` in my own code where basically option is used as a "boolean with more information". The information bit being not used most of the time in the program except on user interaction to have enough information to describe how the condition happened to the end user.

I admit this is really horrible style and promise never to do it again but then maybe that's what happens when one tries to write real world programs with good user experience. In any case this would perfectly and legitimately apply for  `Either.t` and even more so. 

As far as I'm concerned I'm done commenting on this PR and I enjoin @gasche to think about the amount of time collectively lost on this PR while it could have been a routine merge, had the `Option` and `Result` module design been followed. I think this attitude:

> But I'm not interested in proposing code that I actively dislike as my own.

is fine when new things are being requested one disgrees with. But not when these things are requested for the consistency of the stdlib especially with a design that occured just a year ago and is not perceived by many others as being flawed.

The functions still missing for a consistent design are mentioned in [this message](https://github.com/ocaml/ocaml/pull/9066#issuecomment-545868207).

**Comment by gasche** [comment]
I thought a bit about the "dataful boolean" view myself, wondering if would justify those operators. #8782 (closed) proposed conversion functions between boleans and options, but it restricted them to `unit option` which are actually isomorphic to booleans. The criticism we made at the time is that there was no canonical association from `Some|None` to `true|false` or `false|true`. There is no ambiguity with `is_{left,right}` about which branch returns `true`, but this shows that they are different sort of operations -- they are not about working with dataful booleans.

In any case, my intuition is that `result` is more appropriate than `Either.t` for dataful booleans (it's clear which case is "`true` with more information" and which is "`false` with more information"), but maybe concrete use-cases could provide further insights there.

**Comment by dbuenzli** [comment]
> In any case, my intuition is that `result` is more appropriate than `Either.t` for dataful booleans

No. You may simply have a mutually exclusive condition none of which qualifies as an `Error` as far as your code is concerned.

**Comment by gasche** [comment]
I have been thinking again about this PR. (What follows is an intricate argument for why you should do things my way :- ).

I believe that for the stdlib it is better to err on the side of not having enough functions rather than having too many functions, because this leaves the door open for improvement by consistent extension in the future (whereas improvement by removal is difficult or impossible). If there is no consensus on a given function which is part of a bigger module, the safe choice would be to omit the function and consider the rest of the module, unless of course the argument can be made that the functions are essential to the module and it does not provide value without them.

**Comment by dbuenzli** [comment]
I think I gave enough arguments in the whole discussion I'm not spending any minute anymore on that PR. You are making the stdlib inconsistent with recent additions and don't have my approval but then I don't have any power here.

**Comment by gasche** [comment]
I'm still right about `is_{left,right}` being API garbage, and people who want it are still wrong (*especially* @dbuenzli) but I decided that I don't really care anymore if the API is wrong for me and right for basically everyone else, so I added the two functions back in.

**Comment by gasche** [comment]
(I also removed `flip : ('a, 'b) Either.t -> ('b, 'a) Either.t` because @trefis was mean to it and it didn't want to come back.)

File: stdlib/either.ml
Comment (xvw) [inline stdlib/either.ml]: Is there any reason to have `(**)`?

Comment (gasche) [reply stdlib/either.ml]: There is one in the `.mli`: otherwise the following documentation comment is attached to the constructor `Right`, not to the whole type. Here there is one in the `.ml` file, and I don't think there is a reason there, that's a good point. It must come from copy-pasting from the interface file.

Comment (gasche) [reply stdlib/either.ml]: I fixed it and rebased, thanks!

**Review by xvw — APPROVED** [review]
From my point of view, the API is more than sufficient. (I'm one of those who find `is_left` `is_right` useful so I'm happy to see them in PR).

**Comment by nojb** [comment]
This PR is stalled even though the discussion seems to have (kind-of) converged on the `is_*` issue.

As far as I can see the only other outstanding point is whether to have `find_left : ('a, 'b) t -> 'a option` and/or `value_left : ('a, 'b) -> default:'a -> 'a`.

Personally I wouldn't be shocked to have both. By analogy with `Result`, I would use the names `{left,right}_to_option` instead of `find_{left,right}`.

Opinions @gasche @yallop @dbuenzli ?

**Comment by yallop** [comment]
@nojb: since you asked, I'm not especially keen on `left_to_option`, since the proposed function converts both `Left v` and `Right v` to `option`.

**Comment by gasche** [comment]
I'm getting a bit tired of waiting for no decision on this PR. @dbuenzli has [posted a clear comment](https://github.com/ocaml/ocaml/pull/9066#issuecomment-545868207) on the fact that he would like extra functions to be added to the module (`get_{left,right}` and `value_{left,right}`; in addition to `is_{left,right}` that I added, and to `find_{left,right}` that exposes the same functionality through composition). Could *other people* please say what they think is required for this PR to be acceptable, or say clearly that we should close the PR and forget about it?

(I still find it exceptional that we would stall stdlib PRs on the argument that some functions are *missing*, given how easy it is to add new functions and how hard it is to remove existing functions.)

**Comment by yallop** [comment]
@gasche: I'm in favour of merging this PR either as is, or with the extra functions that @dbuenzli proposes.

**Comment by dbuenzli** [comment]
Despite what I said above I did look at again the PR :-) I'm happy about the addition of  `is_*` functions and I think `find_{left,right}` can be enough for now.

**Review by nojb — APPROVED** [review]
I am formally approving based on the discussion and a quick review of the code. Good to go after rebasing. Thanks @gasche!

**Comment by gasche** [comment]
Great! Thanks everyone for the surprising convergence speed. I rebased on current trunk.

File: Changes
Comment (yallop) [inline Changes]: This seems to remove several existing entries from `Changes`.  Is that a mistake?

File: stdlib/list.ml
Comment (yallop) [inline stdlib/list.ml]: It would be a little clearer to call `yes` and `no` `left` and `right`.

**Review by yallop — CHANGES_REQUESTED** [review]

File: Changes
Comment (gasche) [reply Changes]: Yes, I did a mistake when rebasing, thanks! I will fix this and rebase. (I can also change `yes` in `left`; I kept the naming scheme of `partition` to emphasize the similarities between the two implementations.))

File: stdlib/list.ml
Comment (gasche) [reply stdlib/list.ml]: (done)

**Review by yallop — APPROVED** [review]

**Comment by yallop** [comment]
There are a few test failures, which are fixed by https://github.com/gasche/ocaml/pull/7.

**Comment by gasche** [comment]
@yallop I suspect you are hinting at the fact that I should wrap up this PR and merge. Thanks for the help with the test failures! I suspect there may be more to come (I would expect that some flambda-only tests are in the way, and that I should probably remove them rather than add friction to every stdlib PR), and I'm happy to get it done before the end of the week. (Feel free to ping again if I forget.)

**Comment by yallop** [comment]
@gasche:
> (Feel free to ping again if I forget.)

Thanks!  Please consider this comment that ping.

It seems that there's just one location-related testsuite failure remaining, though I suppose rebasing might reveal more.

**Comment by gasche** [comment]
I submitted #9871 as a more long-term solution to the testsuite noise. I would like to wait for that short PR to follow its natural course, then rebase the present one on top of the final solution.

**Comment by gasche** [comment]
(Thanks for the merge!)

**Pull request merged**
